bindings/alicloud/dingtalk/webhook/webhook.go

```diff
@@ -0,0 +1,218 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation and Dapr Contributors.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	webhookContentType       = "application/json"
+	defaultHTTPClientTimeout = time.Second * 30
+)
+
+type DingTalkWebhook struct {
+	logger     logger.Logger
+	settings   Settings
+	httpClient *http.Client
+}
+
+type webhookResult struct {
+	ErrCode int    `json:"errcode"`
+	ErrMsg  string `json:"errmsg"`
+}
+
+type outgoingWebhook struct {
+	handler func(*bindings.ReadResponse) ([]byte, error)
+}
+
+var webhooks sync.Map //nolint:gochecknoglobals
+
+func NewDingTalkWebhook(l logger.Logger) *DingTalkWebhook {
+	// See guidance on proper HTTP client settings here:
+	// https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779
+	dialer := &net.Dialer{
+		Timeout: 5 * time.Second,
+	}
+	var netTransport = &http.Transport{
+		Dial:                dialer.Dial,
+		TLSHandshakeTimeout: 5 * time.Second,
+	}
+	httpClient := &http.Client{
+		Timeout:   defaultHTTPClientTimeout,
+		Transport: netTransport,
+	}
+
+	return &DingTalkWebhook{ //nolint:exhaustivestruct
+		logger:     l,
+		httpClient: httpClient, //nolint:exhaustivestruct
+	}
+}
+
+// Init performs metadata parsing
+func (t *DingTalkWebhook) Init(metadata bindings.Metadata) error {
+	var err error
+	if err = t.settings.Decode(metadata.Properties); err != nil {
+		return fmt.Errorf("dingtalk configuration error: %w", err)
+	}
+	if err = t.settings.Validate(); err != nil {
+		return fmt.Errorf("dingtalk configuration error: %w", err)
+	}
+
+	return nil
+}
+
+// Read triggers the outgoing webhook, not yet production ready
+func (t *DingTalkWebhook) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
+	t.logger.Debugf("dingtalk webhook: start read input binding")
+	item := outgoingWebhook{handler: handler}
+
+	if _, loaded := webhooks.LoadOrStore(t.settings.ID, &item); loaded {
```

Comment:
This is probably a case where using the builtin map + mutex gives you more control.  There is no need to store if it already exists. Here it doesn't cause any problem though.